Sprints ui

Avogardo opened the pull request:
Add create project form
Add projects sprints list

Avogardo committed:
add list of sprints

add list of sprints to SingleProjectSidebar

Avogardo committed:
sprint list show only current projects sprints

Fix sprintList query

Avogardo committed:
add route to create project form

Add route to create project component
Add button do the main page

Avogardo committed:
Add form to creating projects

Avogardo committed:
fix lint errors

mstanuch commented on app/imports/api/sprints/server/publications.js:
```diff
@@ -0,0 +1,31 @@
+        return this.ready();
+    }
+    const query = { projectId };
+    limitQueryToUserSprints(this.userId, query);
```
Nice copy and paste! Do you know what this function do?

mstanuch commented on app/imports/components/Hello/Hello.jsx:
```diff
@@ -1,9 +1,23 @@
-Hello.propTypes = {};
+    return (<div>
+        <h1>Hello!</h1>
+        <RaisedButton
```
I think the better place would be project list (sidebar).

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -0,0 +1,72 @@
+
+        const name = this.area.input.value;
+
+        const result = createNewProject(name, [0], [0]);
```
Why do you pass `[0]`?

mstanuch commented on app/imports/components/Projects/SingleProjectSidebar/SingleProjectSidebar.jsx:
```diff
@@ -13,8 +13,27 @@ const getProjectListItems = (projects, onTouchTap) =>
+        />
+    ));
+
+    listSprints[listSprints.length] = (
```
`listSprints.push(...)`

mstanuch commented on app/imports/startup/client/routes.jsx:
```diff
@@ -31,6 +32,7 @@ export default (
     <Route path="/">
         <Route component={MainLayout} onEnter={onlyLoggedIn}>
             <Route path="hello" components={{ main: Hello, drawerContent: ProjectList }} />
+            <Route path="create-project" components={{ main: CreateNewProject, drawerContent: ProjectList }} />
```
In my opinion better path would be `/project/create` or `/project/add`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -0,0 +1,72 @@
+    RaisedButton,
+} from 'material-ui';
+
+class CreateNewProject extends Component {
```
You forgot about adding members and moderators to the project.

mstanuch commented on app/imports/components/Projects/SingleProjectSidebar/SingleProjectSidebar.jsx:
```diff
@@ -26,8 +45,8 @@ const SingleProjectSidebar = (props) => {
             <ListItem
                 primaryText="Sprints"
 
```
Is there any meaning of these blank lines?

Avogardo committed:
Move project create button to the sidebar

Create new project button has been moved to the sidebar
Unnessesary function from sprints publications has been removed
Used arr.push() instead arr[arr.length]
Blank lines from SingleProjectSidebar has been removed
Change route to /project/create

Avogardo committed:
Adding moderators and members when creating project

Avogardo committed:
Fix lint errors

Avogardo committed:
Return throwErrorIfNotAdmin function in createProject method

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
     constructor(props) {
         super(props);
         this.onSubmit = this.onSubmit.bind(this);
+        this.addMods = this.addMods.bind(this);
```
Please, use full name `addModerators`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
             isError: false,
             isResult: false,
+            usersNames: [],
+            usersIds: [],
```
You receive `usersNames` and `usersIds` in `props`.

mstanuch replied:
Instead having two array with separated names and ids, please make one array of objects:
```js
{
 name: ...,
 id: ...,
}
```
When you have object you always know user and his id. Array items can be shuffled.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+            chosenModerators: [],
+            chosenMenbers: [],
+            inputError: false,
+            memberError: false,
```
What do you thing to change `inputError` and `memberError` to an object (or array of objects): 
```js
error: { 
  type: 'inputError',
  message: 'error details'
}
```

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+            usersNames: [],
+            usersIds: [],
+            chosenModerators: [],
+            chosenMenbers: [],
```
Typo.

What do you think about using short names `members` and `moderators`?

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+
+        const names = [];
+        const ids = [];
+        for (let i = 0; i < users.length; i += 1) {
```
```js
users.forEach(user => {
  // function body
});
```

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+        const names = [];
+        const ids = [];
+        for (let i = 0; i < users.length; i += 1) {
+            names.push(users[i].services.github.username);
```
You should get username from user profile, because it's available only when user use GitHub to sing in. 

Use:
`profile.name` instead `services.github.username`.

Avogardo replied:
Profile name is optional, first time I`ve used profile.name and got undefinied
Also profiles names can be duplicated

mstanuch replied:
> Profile name is optional

Yes, `profile.name` is optional, but our specification says that user can change his name, it's the best place to keep it.

> first time I`ve used profile.name and got undefinied

During first sign in Meteor fetches data from GitHub. User won't see `undefined`, because server is fast enough to update this data on both sides.

> Also profiles names can be duplicated

Yes, I know it. But this must be clarified.
In my opinion in this case email would be better to identify user.

Avogardo replied:
So use emails instead of profile.name and service.blablabla?

mstanuch replied:
No, for now use just `profile.name`, we need to clarify it.
If you want to identify user now, make something like `username (email)`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
-        const result = createNewProject(name, [0], [0]);
+        const moderatorsId = [];
+        let index = 0;
+        for (let i = 0; i < chosenModerators.length; i += 1) {
```
Use `forEach()`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+        let index = 0;
+        for (let i = 0; i < chosenModerators.length; i += 1) {
+            index = usersNames.indexOf(chosenModerators[i]);
+            moderatorsId.push(usersIds[index]);
```
If username not found it will push `-1`.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+
+        const membersId = [];
+        index = 0;
+        for (let i = 0; i < chosenMenbers.length; i += 1) {
```
Use `forEach()`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+        const moderatorsId = [];
+        let index = 0;
+        for (let i = 0; i < chosenModerators.length; i += 1) {
+            index = usersNames.indexOf(chosenModerators[i]);
```
`const index = ...`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
 
-        const result = createNewProject(name, [0], [0]);
+        const moderatorsId = [];
+        let index = 0;
```
It looks bad. See comment on line 53.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -35,9 +76,123 @@ class CreateNewProject extends Component {
 
-    renderTextField() {
-        const { isError, isResult } = this.state;
+    addMods(name) {
```
`addMods` suggest plural.

Change function name to `addModerator`

mstanuch commented on app/imports/startup/client/routes.jsx:
```diff
@@ -32,7 +32,7 @@ export default (
         <Route component={MainLayout} onEnter={onlyLoggedIn}>
             <Route path="hello" components={{ main: Hello, drawerContent: ProjectList }} />
-            <Route path="create-project" components={{ main: CreateNewProject, drawerContent: ProjectList }} />
+            <Route path="create" components={{ main: CreateNewProject, drawerContent: ProjectList }} />
```
Path `/project/create` ? :)

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -53,20 +208,36 @@ class CreateNewProject extends Component {
+            {this.renderProjectNameForm()}
+            {this.renderModForm()}
+            {!this.state.isError && this.state.isResult ? '' :
+            <List>
```
What do you think about `Chip` component :) I think that this component is more user friendly for this purpose.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -53,20 +208,36 @@ class CreateNewProject extends Component {
+            {this.renderMemberForm()}
+            {!this.state.isError && this.state.isResult ? '' :
+            <div>
+                <List>
```
As above ;)

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -35,9 +76,123 @@ class CreateNewProject extends Component {
+            this.setState({
+                chosenModerators: moderators,
+            });
+        } else {
```
If you use `type` with value `mod` you should also use other value for `member`. 
Full name are better, because `mod` can be interpreted as `modifier`.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -35,9 +76,123 @@ class CreateNewProject extends Component {
+        this.memberInput.focus();
+    }
+
+    removeRow(e, index, type) {
```
Make two methods `removeModerator` and `removeMember`.

mstanuch commented on app/imports/api/users/server/publications.js:
```diff
@@ -0,0 +1,13 @@
+
+    const options = {
+        fields: {
+            services: 1,
```
Publishing user access tokens, mails and other private data is never good choice!

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -35,9 +76,123 @@ class CreateNewProject extends Component {
+            });
+        }
+
+        this.modInput.setState({ searchText: '' });
```
I'm not sure about this.

mstanuch reviewed (changes requested):
I think that you should make one reusable component to adding users.

KowalColo commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -1,26 +1,67 @@
+        const names = [];
+        const ids = [];
+        for (let i = 0; i < users.length; i += 1) {
+            names.push(users[i].services.github.username);
```
Profile name is optional, first time I`ve used profile.name and got undefinied
Also profiles names can be duplicated

Avogardo committed:
full mod name, user names/id in props, error object, no for loops,

Fix typo
Change functions names
Reduce states
Use forEach instead of for loops

Avogardo committed:
Publication is now safer

Avogardo committed:
States short names and functions full names, Chip

use chip instead list

Avogardo committed:
Fix lint errors, add login button

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
+            members: [],
             inputError: false,
             memberError: false,
+            error: {
```
Choose one:
1. Multiple error vars
2. One error object or array of objects

mstanuch commented on app/imports/api/users/server/publications.js:
```diff
@@ -5,7 +5,7 @@ Meteor.publish('userList', () => {
     const options = {
         fields: {
-            services: 1,
+            'services.github.username': 1,
```
See our [previous discussion](https://github.com/nokia-wroclaw/innovativeproject-retrotool/pull/5#discussion_r107494557), there is new comment ;)

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
-            index = usersNames.indexOf(chosenModerators[i]);
-            moderatorsId.push(usersIds[index]);
-        }
+        let user = '';
```
Remove this line, you can easliy replace with `const`, see below.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
+        let user = '';
+
+        moderators.forEach((moderator) => {
+            user = userList.find(obj => obj.services.github.username === moderator);
```
`const user = ...`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
-            membersId.push(usersIds[index]);
-        }
+        members.forEach((member) => {
+            user = userList.find(obj => obj.services.github.username === member);
```
`const user = ...`

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
-    renderChosenMembers() {
-        const chosen = this.state.chosenMenbers;
-        return chosen.map((name, index) => <ListItem key={`membername${name}`}>
+    rendermembers() {
```
camelCase

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
+
+        const names = [];
+
+        users.forEach(user => names.push(user.services.github.username));
```
`map()`?

`user.services.github.username` is not guaranteed, this line can throw error.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -2,203 +2,223 @@ import React, { Component, PropTypes } from 'react';
+        const users = this.props.userList;
+
+        const names = [];
+        users.forEach(user => names.push(user.services.github.username));
```
`map()`?

mstanuch commented on app/imports/components/layout/index.js:
```diff
@@ -0,0 +1,21 @@
+import { Meteor } from 'meteor/meteor';
```
You do not track any data in this file - unnecessary composer.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -0,0 +1,259 @@
+        this.moderatorInput.focus();
+    }
+
+    addMember(name, names) {
```
Please, identify user by his `id` not by `name`.

mstanuch commented on app/imports/components/Projects/CreateNewProject/CreateNewProject.jsx:
```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    addModerator(name, names) {
```
Please, identify user by his `id` not by `name`.

Avogardo committed:
profile.name instead github.username, removed layout index.js, add datasource with config to AutoComplete

Avogardo committed:
Remove unnecessary functions from project index.js

Avogardo committed:
Add current sprint page
